plugins/out_cloudwatch_logs/cloudwatch_api.c

```diff
@@ -847,6 +847,13 @@ int process_and_send(struct flb_cloudwatch *ctx, const char *input_plugin,
             check = FLB_FALSE;
             found = FLB_FALSE;
 
+            /* Check type of msgpack object */
+            if (root.via.array.ptr[1].type != MSGPACK_OBJECT_MAP) {
+                flb_plg_error(ctx->ins, "Could not find log_key '%s' in record",
+                    ctx->log_key);
+                continue;
+            }
```

Comment:
so right above this code, outside of the log_key if block, I see:

```
/* Get the record/map */
map = root.via.array.ptr[1];
map_size = map.via.map.size;
```

Is that code safe? Is it safe to take the map size before we know its a map? Should we/can we move this code to be inside the log_key if block after this new check?

Reply:
Regarding size, it seems like size is always in the first 32 bits of the struct regardless of whether the struct is a map or a string, array, or map, which is why we were previously seeing that size is the length of the string.

However, I just realized that if logs can be basic types, such as boolean, int, double, then this may be a problem.

```
typedef struct {
    uint32_t size;
    struct msgpack_object* ptr;
} msgpack_object_array;

typedef struct {
    uint32_t size;
    struct msgpack_object_kv* ptr;
} msgpack_object_map;

typedef struct {
    uint32_t size;
    const char* ptr;
} msgpack_object_str;

typedef struct {
    uint32_t size;
    const char* ptr;
} msgpack_object_bin;

typedef struct {
    int8_t type;
    uint32_t size;
    const char* ptr;
} msgpack_object_ext;
```

```
typedef union {
    bool boolean;
    uint64_t u64;
    int64_t  i64;
#if defined(MSGPACK_USE_LEGACY_NAME_AS_FLOAT)
    double   dec; /* obsolete*/
#endif /* MSGPACK_USE_LEGACY_NAME_AS_FLOAT */
    double   f64;
    msgpack_object_array array;
    msgpack_object_map map;
    msgpack_object_str str;
    msgpack_object_bin bin;
    msgpack_object_ext ext;
} msgpack_object_union;

typedef struct msgpack_object {
    msgpack_object_type type;
    msgpack_object_union via;
} msgpack_object;
```